web/realtime/realtime.go

```diff
@@ -0,0 +1,240 @@
+	Payload wsErrorPayload `json:"payload"`
+}
+
+func unauthorized(cmd interface{}) *wsError {
+	return &wsError{
+		Event: "error",
+		Payload: wsErrorPayload{
+			Status: "401 Unauthorized",
+			Code:   "unauthorized",
+			Title:  "The authentication has failed",
+			Source: cmd,
+		},
+	}
+}
+
+func forbidden(cmd *command) *wsError {
+	return &wsError{
+		Event: "error",
+		Payload: wsErrorPayload{
+			Status: "403 Forbidden",
+			Code:   "forbidden",
+			Title:  fmt.Sprintf("The application can't subscribe to %s", cmd.Payload.Type),
+			Source: cmd,
+		},
+	}
+}
+
+func unknownMethod(method string, cmd interface{}) *wsError {
+	return &wsError{
+		Event: "error",
+		Payload: wsErrorPayload{
+			Status: "405 Method Not Allowed",
+			Code:   "method not allowed",
+			Title:  fmt.Sprintf("The %s method is not supported", method),
+			Source: cmd,
+		},
+	}
+}
+
+func missingType(cmd *command) *wsError {
+	return &wsError{
+		Event: "error",
+		Payload: wsErrorPayload{
+			Status: "404 Page Not Found",
+			Code:   "page not found",
+			Title:  "The type parameter is mandatory for SUBSCRIBE",
+			Source: cmd,
+		},
+	}
+}
+
+func readPump(i *instance.Instance, ws *websocket.Conn, ds *realtime.DynamicSubscriber, errc chan *wsError) {
+	var auth map[string]string
+	if err := ws.ReadJSON(&auth); err != nil {
+		errc <- unknownMethod(auth["method"], auth)
+		return
+	}
+	if strings.ToUpper(auth["method"]) != "AUTH" {
+		errc <- unknownMethod(auth["method"], auth)
+		return
+	}
+	if auth["payload"] == "" {
+		errc <- unauthorized(auth)
+		return
+	}
+	pdoc, err := webpermissions.ParseJWT(i, auth["payload"])
+	if err != nil {
+		errc <- unauthorized(auth)
+		return
+	}
+
+	for {
+		cmd := &command{}
+		if err = ws.ReadJSON(cmd); err != nil {
+			if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway) {
+				logger.WithDomain(ds.Domain).Infof("ws error: %s", err)
+			}
+			break
+		}
+
+		if strings.ToUpper(cmd.Method) != "SUBSCRIBE" {
+			errc <- unknownMethod(cmd.Method, cmd)
+			continue
+		}
+		if cmd.Payload.Type == "" {
+			errc <- missingType(cmd)
+			continue
+		}
+		if !pdoc.Permissions.AllowWholeType(permissions.GET, cmd.Payload.Type) {
+			errc <- forbidden(cmd)
+			continue
+		}
+
+		if cmd.Payload.ID == "" {
+			err = ds.Subscribe(cmd.Payload.Type)
+		} else {
+			err = ds.Watch(cmd.Payload.Type, cmd.Payload.ID)
+		}
+		if err != nil {
+			logger.WithDomain(ds.Domain).Warnf("realtime error: %s", err)
+		}
+	}
+}
+
+func ws(c echo.Context) error {
+	instance := middlewares.GetInstance(c)
+
+	ws, err := upgrader.Upgrade(c.Response(), c.Request(), nil)
+	if err != nil {
+		return err
+	}
+	defer ws.Close()
+
+	ws.SetReadLimit(maxMessageSize)
+	if err = ws.SetReadDeadline(time.Now().Add(pongWait)); err != nil {
+		return nil
+	}
+	ws.SetPongHandler(func(string) error {
+		return ws.SetReadDeadline(time.Now().Add(pongWait))
+	})
+
+	ds := realtime.GetHub().Subscriber(instance.Domain)
+	defer ds.Close()
+	errc := make(chan *wsError)
```

Comment:
Adding a buffer-size of `1` would prevent a goroutine (`readPump`) leak in case the reading out of `errc` stops for some reason (client closing its connection causing a `WriteMessage` to fail for example)